include/gc/ExecutionEngine/GPURuntime/GpuOclRuntime.h

```diff
@@ -0,0 +1,321 @@
+    assert(preserveOrder || (waitListLen == 0 && waitList == nullptr));
+  }
+
+  OclContext(const OclContext &) = delete;
+  OclContext &operator=(const OclContext &) = delete;
+
+  void finish();
+
+private:
+  friend OclRuntime;
+  friend OclRuntime::Exports;
+  template <unsigned N> friend struct OclModuleArgs;
+  // Contains the pointers of all non-USM arguments. It's expected, that the
+  // arguments are either USM or CL pointers and most probably are USM, thus,
+  // in most cases, this set will be empty.
+  std::unordered_set<void *> clPtrs;
+
+  void setLastEvent(cl_event event) {
+    lastEvent = event;
+    if (event) {
+      waitListLen = 1;
+      waitList = &lastEvent;
+    } else {
+      waitListLen = 0;
+      waitList = nullptr;
+    }
+  }
+};
+
+// The main function arguments in the following format -
+// https://mlir.llvm.org/docs/TargetLLVMIR/#c-compatible-wrapper-emission.
+// NOTE: The values are not copied, only the pointers are stored!
+// NOTE: This class is mutable and not thread-safe!
+template <unsigned N = 64> struct OclModuleArgs {
+  explicit OclModuleArgs(OclContext &ctx) : ctx(ctx) {}
+  OclModuleArgs(const OclModuleArgs &) = delete;
+  OclModuleArgs &operator=(const OclModuleArgs &) = delete;
+
+  void add(void *&alignedPtr, size_t rank, const int64_t *shape,
+           const int64_t *strides, bool isUsm = true) {
+    add(alignedPtr, alignedPtr, ZERO, rank, shape, strides, isUsm);
+  }
+
+  void add(void *&allocatedPtr, void *&alignedPtr, const int64_t &offset,
+           size_t rank, const int64_t *shape, const int64_t *strides,
+           bool isUsm = true) {
+#ifndef NDEBUG
+    assert(!isUsm || ctx.runtime.isUsm(alignedPtr));
+    // It's recommended to have at least 16-byte alignment
+    assert(reinterpret_cast<std::uintptr_t>(alignedPtr) % 16 == 0);
+#endif
+
+    args.emplace_back(&allocatedPtr);
+    args.emplace_back(&alignedPtr);
+    args.emplace_back(const_cast<int64_t *>(&offset));
+    for (size_t i = 0; i < rank; i++) {
+      args.emplace_back(const_cast<int64_t *>(&shape[i]));
+    }
+    for (size_t i = 0; i < rank; i++) {
+      args.emplace_back(const_cast<int64_t *>(&strides[i]));
+    }
+    if (!isUsm) {
+      ctx.clPtrs.insert(alignedPtr);
+    }
+  }
+
+  template <typename T>
+  void add(T *&alignedPtr, size_t rank, const int64_t *shape,
+           const int64_t *strides, bool isUsm = true) {
+    add(reinterpret_cast<void *&>(alignedPtr), rank, shape, strides, isUsm);
+  }
+
+  template <typename T>
+  void add(T *&allocatedPtr, T *&alignedPtr, const int64_t &offset, size_t rank,
+           const int64_t *shape, const int64_t *strides, bool isUsm = true) {
+    add(reinterpret_cast<void *&>(allocatedPtr),
+        reinterpret_cast<void *&>(alignedPtr), offset, rank, shape, strides,
+        isUsm);
+  }
+
+  void clear() {
+    args.clear();
+    ctx.clPtrs.clear();
+  }
+
+private:
+  friend OclModule;
+  OclContext &ctx;
+  SmallVector<void *, N + 3> args;
```

Comment:
N could be used by the user code, when the number of args is known in compile time (see the unit test, for example). It allows to allocate the array on stack. +3 is because we add 3 additional args.

> size of the vector to be `66`, which may not align in the memory as good as `64`

This is 66 elements of type void*, but not 66 bytes.

> BTW, do we have a "well-motivated reason" to set our own N?

As I remember, the default implementation uses 64 bytes by default. This is 8 args, that is not enough in out case.

Reply:
In most cases, I think, no. The default value should be sufficient in most cases, but if not, it could be increased, decreased, or set to 0 in order to disable on stack allocation. It gives more flexibility with the memory allocation for the arguments.